source/Applications/Advanced/TransformViaArucoMarker/TransformViaArucoMarker.cpp

```diff
@@ -0,0 +1,308 @@
+        return { slope, intercept };
+    };
+
+    cv::Mat pointCloudToBGR(const Zivid::PointCloud &pointCloud)
+    {
+        auto rgb = cv::Mat(pointCloud.height(), pointCloud.width(), CV_8UC4); // NOLINT(hicpp-signed-bitwise)
+        pointCloud.copyData(
+            reinterpret_cast<Zivid::ColorRGBA *>(rgb.data)); // NOLINT(cppcoreguidelines-pro-type-reinterpret-cast)
+        auto bgr = cv::Mat(pointCloud.height(), pointCloud.width(), CV_8UC4); // NOLINT(hicpp-signed-bitwise)
+        cv::cvtColor(rgb, bgr, cv::COLOR_RGBA2BGR);
+
+        return bgr;
+    }
+
+    std::vector<cv::Point2f> estimate2DMarkerCenters(const aruconano::Marker &markers)
+    {
+        if(markers.empty())
+        {
+            std::cout << "No markers detected";
+            return {};
+        }
+
+        std::vector<cv::Point2f> markerCenters2D;
+        markerCenters2D.reserve(markers.size());
+
+        auto p11 = markers[0];
+        auto p12 = markers[1];
+        auto p21 = markers[3];
+        auto p22 = markers[2];
+
+        const auto backDiagonal{ fitLine(p11, p22) };
+        const auto forwardDiagonal{ fitLine(p21, p12) };
+        const auto xCoord =
+            (forwardDiagonal.intercept - backDiagonal.intercept) / (backDiagonal.slope - forwardDiagonal.slope);
+        const auto yCoord = backDiagonal.slope * xCoord + backDiagonal.intercept;
+        markerCenters2D.emplace_back(cv::Point2f(xCoord, yCoord));
+
+        return markerCenters2D;
+    }
+
+    std::vector<cv::Point2f> get2DMarkerCorners(const aruconano::Marker &markers)
+    {
+        if(markers.empty())
+        {
+            std::cout << "No markers detected";
+            return {};
+        }
+
+        std::vector<cv::Point2f> markerCorners2D;
+        markerCorners2D.reserve(markers.size() * 4);
+
+        markerCorners2D.emplace_back(markers[0]);
+        markerCorners2D.emplace_back(markers[1]);
+        markerCorners2D.emplace_back(markers[2]);
+        markerCorners2D.emplace_back(markers[3]);
+
+        return markerCorners2D;
+    }
+
+    std::vector<cv::Point3d> estimate3DMarkerPoints(const Zivid::PointCloud &pointCloud,
+                                                    const std::vector<cv::Point2f> &markerPoints2D)
+    {
+        if(markerPoints2D.empty())
+        {
+            return {};
+        }
+
+        int w = pointCloud.width();
+        int h = pointCloud.height();
+        const auto points = pointCloud.copyPointsXYZ();
+
+        std::vector<cv::Point3d> markerPoints3D;
+        markerPoints3D.reserve(markerPoints2D.size());
+        for(const auto &point2D : markerPoints2D)
+        {
+            float x = point2D.x;
+            float y = point2D.y;
+
+            int x1 = clamp(static_cast<int>(std::floor(x)), 0, w - 1);
+            int x2 = clamp(static_cast<int>(std::ceil(x)), 0, w - 1);
+            int y1 = clamp(static_cast<int>(std::floor(y)), 0, h - 1);
+            int y2 = clamp(static_cast<int>(std::ceil(y)), 0, h - 1);
```

Comment:
What does this do? Add a comment or rename variables so that it is more clear?

Reply:
The functions clamps/clips a number (here x and y) between the interval 0 and w-1 or h-1. If it is within, the output is x, if its below, the output is 0 and if its above the output is w-1 or h-1. 

From C++17 this is an internal function in std, but since we require that the SDK is functional with C++11, I needed to create it myself. https://en.cppreference.com/w/cpp/algorithm/clamp

Reply:
I understand what the function does. But it is not clear why it is needed in the algorithm. The answer has to do with the bilinear interpolation. Adding comments in the code to explain the algorithm a bit will solve this.

Reply:
And by changing the variable names I meant that e.g. xRounded or something like that would say more than x1.

Reply:
done